File: benches/lang-comparison/python/http_types.py
# HTTP types with error handling
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

class HttpErrorKind(Enum):
    BAD_REQUEST = "BadRequest"
    NOT_FOUND = "NotFound"
    INTERNAL = "Internal"

@dataclass
class HttpError(Exception):
    kind: HttpErrorKind
    message: str

@dataclass
class Header:
    name: str
    value: str

@dataclass
class Request:
    method: str
    path: str
    headers: list[Header] = field(default_factory=list)
    body: str = ""

@dataclass
class Response:
    status: int
    headers: list[Header] = field(default_factory=list)
    body: str = ""

def parse_method(s: str) -> str:
    if s in ("GET", "POST", "PUT", "DELETE"):
        return s
    raise HttpError(HttpErrorKind.BAD_REQUEST, "invalid method")

def route(req: Request) -> Response:
    if req.path == "/":
        return Response(status=200, body="OK")
    elif req.path == "/health":
        return Response(status=200, body="healthy")
    else:
        return Response(status=404, body="not found")

def format_response(res: Response) -> str:
    return f"HTTP/1.1 {res.status} {res.body}"

def main():
    req = Request(method="GET", path="/")
    res = route(req)
    print(format_response(res))

if __name__ == "__main__":
    main()

File: benches/lang-comparison/go/http_types.go
// HTTP types with error handling
package main

import (
	"errors"
	"fmt"
)

type HttpError struct {
	Kind    string
	Message string
}

func (e *HttpError) Error() string {
	return fmt.Sprintf("%s: %s", e.Kind, e.Message)
}

type Header struct {
	Name  string
	Value string
}

type Request struct {
	Method  string
	Path    string
	Headers []Header
	Body    string
}

type Response struct {
	Status  int
	Headers []Header
	Body    string
}

func parseMethod(s string) (string, error) {
	switch s {
	case "GET", "POST", "PUT", "DELETE":
		return s, nil
	default:
		return "", errors.New("invalid method")
	}
}

func route(req *Request) Response {
	switch req.Path {
	case "/":
		return Response{Status: 200, Headers: nil, Body: "OK"}
	case "/health":
		return Response{Status: 200, Headers: nil, Body: "healthy"}
	default:
		return Response{Status: 404, Headers: nil, Body: "not found"}
	}
}

func formatResponse(res Response) string {
	return fmt.Sprintf("HTTP/1.1 %d %s", res.Status, res.Body)
}

func main() {
	req := &Request{
		Method:  "GET",
		Path:    "/",
		Headers: nil,
		Body:    "",
	}
	res := route(req)
	fmt.Println(formatResponse(res))
}
